### PR TITLE
Added integration test for Relay Maintainer

### PR DESCRIPTION
**Pull request by tomaszslabon**
This pull request contains an integration test for `Relay Maintainer`'s `relay`:
 - `Relay` uses "local" instances of BTC and ETH nodes for easy testing
 - `Relay` is parametrised to have a shorter `difficultyEpochDuration` and `pushingSleepTime` for faster execution
 - Some changes beyond test implementation:
    - Function `startRelay` added
    - Function `splitBatch` is no longer a free function, but `Relay`'s method
    - `Relay` has fields `difficultyEpochDuration` and `pushingSleepTime`

### CODE REVIEW

File: relay/pkg/header/integration_test.go [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+	// Set the best known digest to header 7, so the relay will be pulling headers
+	// starting with the 8th header
+	localChain.SetBestKnownDigest(to32Bytes(7))
+
+	// Start the relay with difficulty changing at every eighth header and
+	// a shortened pushing loop sleep time
+	const (
+		testDifficultyEpochDuration = 8
+		testRelayPushingSleepTime   = 200 * time.Millisecond
+	)
```
Comment (lukasz-zimnoch) [inline relay/pkg/header/integration_test.go:68]: Can we put those consts at the top of the file and document them? This way when someone reads this file he will be aware of them from the beginning.

File: relay/pkg/header/integration_test.go [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+	startRelay(
+		ctx,
+		btcChain,
+		localChain,
+		testDifficultyEpochDuration,
+		relayPullingSleepTime,
+		testRelayPushingSleepTime,
+		&mockObserver{},
+	)
+	time.Sleep(10 * time.Millisecond)
```
Comment (lukasz-zimnoch) [inline relay/pkg/header/integration_test.go:78]: Worth documenting why are we sleeping here.

File: relay/pkg/header/integration_test.go [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+		t.Errorf(
+			"unexpected mark new heaviest event:\n"+
+				"expected: [%+v]\n"+
+				"actual:   [%+v]\n",
+			expectedMarkNewHeaviestEvent,
+			actualMarkNewHeaviestEvent,
+		)
+	}
+
+	localChain.SetBestKnownDigest(to32Bytes(12))
```
Comment (lukasz-zimnoch) [inline relay/pkg/header/integration_test.go:147]: Why are we setting the best digest manually? Isn't it set automatically on `MarkNewHeaviest` local call?

File: relay/pkg/header/integration_test.go [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+	time.Sleep(testRelayPushingSleepTime)
```
Comment (lukasz-zimnoch) [inline relay/pkg/header/integration_test.go:148]: Worth documenting why are we sleeping.

File: relay/pkg/header/integration_test.go [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+				"expected: [%+v]\n"+
+				"actual:   [%+v]\n",
+			expectedAddHeadersWithRetargetEvent,
+			actualAddHeadersWithRetargetEvent,
+		)
+	}
+
+	markNewHeaviestEvents = localChain.MarkNewHeaviestEvents()
+
+	expectedEventsCount = 2
```
Comment (lukasz-zimnoch) [inline relay/pkg/header/integration_test.go:217]: Worth putting a comment on why we have two events here. May be not obvious for someone not familiar with relay logic.

File: relay/pkg/header/integration_test.go [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+		)
+	}
+
+	expectedMarkNewHeaviestEvent = &chainlocal.MarkNewHeaviestEvent{
+		AncestorDigest:    to32Bytes(12),
+		CurrentBestHeader: toBytes(12),
+		NewBestHeader:     toBytes(17),
+		Limit:             big.NewInt(6),
+	}
+	actualMarkNewHeaviestEvent = markNewHeaviestEvents[1]
```
Comment (lukasz-zimnoch) [inline relay/pkg/header/integration_test.go:235]: Worth documenting why are we taking event with index `1`

File: relay/pkg/header/integration_test.go [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+		t.Errorf(
+			"unexpected mark new heaviest event:\n"+
+				"expected: [%+v]\n"+
+				"actual:   [%+v]\n",
+			expectedMarkNewHeaviestEvent,
+			actualMarkNewHeaviestEvent,
+		)
+	}
+
+	localChain.SetBestKnownDigest(to32Bytes(17))
```
Comment (lukasz-zimnoch) [inline relay/pkg/header/integration_test.go:246]: Same as above regarding setting this manually. Is it needed?

File: relay/pkg/header/integration_test.go [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+	time.Sleep(testRelayPushingSleepTime + headerTimeout)
```
Comment (lukasz-zimnoch) [inline relay/pkg/header/integration_test.go:247]: Worth documenting why are we waiting this specific time.

File: relay/pkg/header/integration_test.go [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+
+	//************ Verify processing of the third batch ************
+	addHeadersEvents = localChain.AddHeadersEvents()
+
+	expectedEventsCount = 2
```
Comment (lukasz-zimnoch) [inline relay/pkg/header/integration_test.go:252]: Worth explaining why we expect two events here.

File: relay/pkg/header/integration_test.go [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+			expectedEventsCount,
+			actualEventsCount,
+		)
+	}
+
+	expectedAddHeadersEvent = &chainlocal.AddHeadersEvent{
+		AnchorHeader: toBytes(17),
+		Headers:      toBytes(18, 19, 20),
+	}
+	actualAddHeadersEvent = addHeadersEvents[1]
```
Comment (lukasz-zimnoch) [inline relay/pkg/header/integration_test.go:268]: And here why are we taking the event with index `1`

File: relay/pkg/header/integration_test.go [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+				"expected: [%+v]\n"+
+				"actual:   [%+v]\n",
+			expectedAddHeadersEvent,
+			actualAddHeadersEvent,
+		)
+	}
+
+	addHeadersWithRetargetEvents = localChain.AddHeadersWithRetargetEvents()
+
+	expectedEventsCount = 2
```
Comment (lukasz-zimnoch) [inline relay/pkg/header/integration_test.go:182]: Please put a comment about why are we expecting two events here.

File: relay/pkg/header/integration_test.go [hunk 12]
```diff
@@ -0,0 +1,292 @@
...
+		)
+	}
+
+	expectedAddHeadersWithRetargetEvent = &chainlocal.AddHeadersWithRetargetEvent{
+		OldPeriodStartHeader: toBytes(8),
+		OldPeriodEndHeader:   toBytes(15),
+		Headers:              toBytes(16, 17),
+	}
+
+	actualAddHeadersWithRetargetEvent = addHeadersWithRetargetEvents[1]
```
Comment (lukasz-zimnoch) [inline relay/pkg/header/integration_test.go:200]: Please document why are we taking event with index `1`

File: relay/pkg/header/integration_test.go [hunk 13]
```diff
@@ -0,0 +1,292 @@
...
+		btcChain,
+		localChain,
+		testDifficultyEpochDuration,
+		relayPullingSleepTime,
+		testRelayPushingSleepTime,
+		&mockObserver{},
+	)
+	time.Sleep(10 * time.Millisecond)
+
+	//************ Verify processing of the first batch ************
```
Comment (lukasz-zimnoch) [inline relay/pkg/header/integration_test.go:80]: Can we add a short summary of what results are we expecting here in a human-friendly form?

File: relay/pkg/header/integration_test.go [hunk 14]
```diff
@@ -0,0 +1,292 @@
...
+				"actual:   [%+v]\n",
+			expectedMarkNewHeaviestEvent,
+			actualMarkNewHeaviestEvent,
+		)
+	}
+
+	localChain.SetBestKnownDigest(to32Bytes(12))
+	time.Sleep(testRelayPushingSleepTime)
+
+	//************ Verify processing of the second batch ************
```
Comment (lukasz-zimnoch) [inline relay/pkg/header/integration_test.go:150]: Can we add a short summary of what results are we expecting here in a human-friendly form?

File: relay/pkg/header/integration_test.go [hunk 15]
```diff
@@ -0,0 +1,292 @@
...
+				"actual:   [%+v]\n",
+			expectedMarkNewHeaviestEvent,
+			actualMarkNewHeaviestEvent,
+		)
+	}
+
+	localChain.SetBestKnownDigest(to32Bytes(17))
+	time.Sleep(testRelayPushingSleepTime + headerTimeout)
+
+	//************ Verify processing of the third batch ************
```
Comment (lukasz-zimnoch) [inline relay/pkg/header/integration_test.go:249]: Can we add a short summary of what results are we expecting here in a human-friendly form?

File: relay/pkg/header/integration_test.go [hunk 16]
```diff
@@ -0,0 +1,292 @@
...
+	btcChain := bc.(*btc.LocalChain)
+
+	lc, err := chainlocal.Connect()
+	if err != nil {
+		t.Fatal(err)
+	}
+	localChain := lc.(*chainlocal.Chain)
+
+	headers := []*btc.Header{
+		{Hash: to32Bytes(0), Height: 0, PrevHash: to32Bytes(0), Raw: toBytes(0)},
```
Comment (lukasz-zimnoch) [inline relay/pkg/header/integration_test.go:31]: Can we clarify that we simulate the "zero" batch is already relayed on the host chain?

File: relay/pkg/header/integration_test.go [hunk 17]
```diff
@@ -0,0 +1,292 @@
...
+				"expected: [%+v]\n"+
+				"actual:   [%+v]\n",
+			expectedAddHeadersEvent,
+			actualAddHeadersEvent,
+		)
+	}
+
+	// no change in MarkNewHeaviestEvents as there were only 3 headers in the batch
+	markNewHeaviestEvents = localChain.MarkNewHeaviestEvents()
+	expectedEventsCount = 2
```
Comment (lukasz-zimnoch) [inline relay/pkg/header/integration_test.go:281]: Worth explaining why we expect two events here.

**Review by lukasz-zimnoch — COMMENTED** [review]
Looks good! Just minor comments regarding docs and some questions to clarify.

File: relay/pkg/header/integration_test.go (hunk 3 above)
Comment (lukasz-zimnoch) [reply relay/pkg/header/integration_test.go:157]: Ok got it. Let's just make it clear why we need to do it here by adding an explanatory comment.

File: relay/pkg/header/integration_test.go (hunk 7 above)
Comment (lukasz-zimnoch) [reply relay/pkg/header/integration_test.go:259]: See https://github.com/keep-network/tbtc/pull/777#discussion_r607665862

File: relay/pkg/header/integration_test.go (hunk 3 above)
Comment (tomaszslabon) [reply relay/pkg/header/integration_test.go:160]: Ok, I've added some explanation.

**Review by lukasz-zimnoch — APPROVED** [review]

**Pull request merged**